internal/kuberesource/mutators.go

```diff
@@ -95,17 +85,84 @@ const (
 	ServiceMeshDisabled serviceMeshMode = "service-mesh-disabled"
 )
 
-// AddServiceMesh adds a service mesh proxy to the resource.
+// AddServiceMesh adds a service mesh proxy to the resource with the proxy
+// configuration given in the object annotations.
 //
 // If the resource does not contain a PodSpec, this function does nothing.
-// This function is not idempotent.
+// This function is idempotent.
 func AddServiceMesh(
 	resource any,
 	serviceMeshProxy *applycorev1.ContainerApplyConfiguration,
-) any {
-	return MapPodSpec(resource, func(spec *applycorev1.PodSpecApplyConfiguration) *applycorev1.PodSpecApplyConfiguration {
+) (res any, retErr error) {
+	res = MapPodSpecWithMeta(resource, func(meta *applymetav1.ObjectMetaApplyConfiguration, spec *applycorev1.PodSpecApplyConfiguration) *applycorev1.PodSpecApplyConfiguration {
+		if spec.RuntimeClassName == nil || !strings.HasPrefix(*spec.RuntimeClassName, "contrast-cc") {
+			return spec
+		}
+
+		ingressConfig, ingressOk := meta.Annotations[smIngressConfigAnnotationKey]
+		egressConfig, egressOk := meta.Annotations[smEgressConfigAnnotationKey]
+		portAnnotation, portOk := meta.Annotations[smAdminInterfaceAnnotationKey]
+
+		// Don't change anything if automatic service mesh injection isn't enabled.
+		if !ingressOk && !egressOk && !portOk {
+			return spec
+		}
+
+		// Remove already existing init containers with unique service mesh name.
+		spec.InitContainers = slices.DeleteFunc(spec.InitContainers, func(c applycorev1.ContainerApplyConfiguration) bool {
+			return *c.Name == *serviceMeshProxy.Name
+		})
+
+		retErr = ensureVolumeExists(spec, *serviceMeshProxy.VolumeMounts[0].Name)
+		if retErr != nil {
+			return nil
+		}
+
+		if portAnnotation != "" {
+			port, err := strconv.Atoi(portAnnotation)
+			if retErr != nil {
+				retErr = fmt.Errorf("parsing service mesh admin interface port: %w", err)
+				return nil
+			}
+
+			ingressConfig += fmt.Sprintf("##admin#%s#true", portAnnotation)
+			serviceMeshProxy.
+				WithEnv(NewEnvVar("EDG_ADMIN_PORT", portAnnotation)).
+				WithPorts(
+					ContainerPort().
+						WithName("admin-interface").
```

Comment:
Maybe more specific? The container might have its own admin interface.

```suggestion
						WithName("contrast-admin-interface").
```